openmeter/billing/service/lineservice/manualusagebasedline.go

```diff
@@ -0,0 +1,137 @@
+package lineservice
+
+import (
+	"context"
+	"time"
+
+	"github.com/samber/lo"
+
+	billingentity "github.com/openmeterio/openmeter/openmeter/billing/entity"
+	"github.com/openmeterio/openmeter/openmeter/productcatalog/plan"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+var _ Line = manualUsageBasedLine{}
+
+type manualUsageBasedLine struct {
+	lineBase
+}
+
+func (l manualUsageBasedLine) PrepareForCreate(context.Context) (Line, error) {
+	l.line.Period = l.line.Period.Truncate(billingentity.DefaultMeterResolution)
+
+	return l, nil
+}
+
+func (l manualUsageBasedLine) Validate(ctx context.Context, targetInvoice *billingentity.Invoice) error {
+	if _, err := l.service.resolveFeatureMeter(ctx, l.line.Namespace, l.line.ManualUsageBased.FeatureKey); err != nil {
+		return err
+	}
+
+	if err := l.lineBase.Validate(ctx, targetInvoice); err != nil {
+		return err
+	}
+
+	if len(targetInvoice.Customer.Subjects) == 0 {
+		return billingentity.ValidationError{
+			Err: billingentity.ErrInvoiceCreateUBPLineCustomerHasNoSubjects,
+		}
+	}
+
+	if l.line.LineBase.Period.Truncate(billingentity.DefaultMeterResolution).IsEmpty() {
+		return billingentity.ValidationError{
+			Err: billingentity.ErrInvoiceCreateUBPLinePeriodIsEmpty,
+		}
+	}
+
+	return nil
+}
+
+func (l manualUsageBasedLine) CanBeInvoicedAsOf(ctx context.Context, asof time.Time) (*billingentity.Period, error) {
+	if l.line.ManualUsageBased.Price.Type() == plan.TieredPriceType {
+		tiered, err := l.line.ManualUsageBased.Price.AsTiered()
+		if err != nil {
+			return nil, err
+		}
+
+		if tiered.Mode == plan.GraduatedTieredPrice {
+			// Graduated tiers are only billable if we have all the data acquired, as otherwise
+			// we might overcharge the customer (if we are at tier bundaries)
+			if !asof.Before(l.line.InvoiceAt) {
+				return &l.line.Period, nil
+			}
+			return nil, nil
+		}
+	}
+
+	meterAndFactory, err := l.service.resolveFeatureMeter(ctx, l.line.Namespace, l.line.ManualUsageBased.FeatureKey)
+	if err != nil {
+		return nil, err
+	}
+
+	meter := meterAndFactory.meter
+
+	asOfTruncated := asof.Truncate(billingentity.DefaultMeterResolution)
+
+	switch meter.Aggregation {
+	case models.MeterAggregationSum, models.MeterAggregationCount,
+		models.MeterAggregationMax, models.MeterAggregationUniqueCount:
```

Comment:
I'm not sure Max works here. Max of period chunk is different then max of whole period

Reply:
Yes, sorry, I was in a bit of rush for the meters, I have updated the docs and the algorithm.